contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -143,39 +175,107 @@ message NamespaceOption {
+- NOT Supported: If there is only one gid and one uid mapping with all the values as `0`,for example `host_id 0, container_id 0, size 0`
+- Supported but disabled: If there is only one gid and one uid mapping. But the size is `4294967295`(max value for uint32) i.e `host_id 0, container_id 0, size 4294967295`
+- Supported and enabled: Host {U/G}ID is not `0` and size is anything between `0` and `4294967295`                                
+
+### Pod admission at Kubelet
+A pod admit handler will be introduced which will fail the pod admission at Kubelet if any of the following conditions is met:
+- If pod spec has non-nil `hostUserNamespace` value AND the container runtime is sandbox type (does not support usernamespace remapping)
+- If pod spec has `hostuserNamespace: false` AND usernamespace remapping is NOT enabled at container runtime
+- If pod has **RunAsUser** or **RunAsGroup** AND  usernamespace remapping is enabled at runtime AND no mapping is found in idmappings corresponding to the **RunAsUser** or **RunAsGroup**
+- If pod has **FsGroup** or **SupplementalGroups** AND  usernamespace remapping is enabled at runtime AND mapping is found in idmappings corresponding to the **FsGroup** or **SupplementalGroups** where container id is not equal to host id i.e mapping is not 1:1
+
+
+## Runtime(runc based) Support
+
 - Docker: Here is the [user-namespace documentation](https://docs.docker.com/engine/security/userns-remap/) and this is the [implementation PR](https://github.com/moby/moby/pull/12648)
     - Concerns:
 Docker API does not provide user-namespace mapping. Therefore to handle `GetRuntimeConfigInfo` API, changes will be done in `dockershim` to read system files, `/etc/subuid` and `/etc/subgid`, for figuring out default user-namespace mapping. `/info` api will be used to figure out if user-namespace is enabled and `Docker Root Dir` will be used to figure out host uid mapped to the uid `0` in container. eg. `Docker Root Dir: /var/lib/docker/2131616.2131616` this shows host uid `2131616` will be mapped to uid `0`
-- CRI-O: https://github.com/kubernetes-incubator/cri-o/pull/1519
+- CRI-O: CRI-O supports usernamespace remapping. [Here is the PR](https://github.com/kubernetes-sigs/cri-o/pull/1941) for getting supplemental groups(fsgroups) working in namespaced environnment.
 - Containerd: https://github.com/containerd/containerd/blob/129167132c5e0dbd1b031badae201a432d1bd681/container_opts_unix.go#L149
 
-### Implementation Roadmap
-#### Phase 1: Support in Kubelet, Alpha, [Target: Kubernetes v1.11]
-- Add feature gate `NodeUserNamespace`, disabled by default
+## Sandbox type runtimes
+`HostUserNamespace` will work only with runc based runtimes. Techniques such as node selectors will have to be used in pods(that are using `HostuserNamespace`) to correctly assign runtimes/nodes. If such a pod unexpectedly lands on a sandbox type runtime node, pod admit handler will reject it.
```

Comment:
I think a node has to support both sandboxed pods and non-sandboxed pods in most cases, but I agreed that pod admit handler still can reject a pod if it request both HostUserNamespace & Sandbox=true.